---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [32]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle
import matplotlib.pyplot as plt

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [3]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [ ]:
def graph_identification():
    
    '''
    def plotGraph(Graph):
        degrees = Graph.degree()
        degree_values = sorted(set(degrees.values()))
        histogram = [list(degrees.values()).count(i)/float(nx.number_of_nodes(Graph)) for i in degree_values]
        plt.bar(degree_values, histogram)
        plt.show()

        print(nx.average_clustering(G2), nx.average_clustering(G3), nx.average_clustering(G5))
        print(nx.average_shortest_path_length(G2), nx.average_shortest_path_length(G3), nx.average_shortest_path_length(G5))
    '''
    
    answer = ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']
    
    return answer

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [121]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

### Scratch Paper

In [3]:
# df = pd.DataFrame(index=G.nodes())

# #Feature Extraction
# df['Department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
# df['ManagementSalary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
# df['Clustering'] = pd.Series(nx.clustering(G))
# df['Degree'] = pd.Series(G.degree())
# df['Deg_Cent'] = pd.Series(nx.degree_centrality(G))
# df['Clo_Cent'] = pd.Series(nx.closeness_centrality(G))
# df['Btn_Cent'] = pd.Series(nx.betweenness_centrality(G))
# df['PR_Scaled'] = pd.Series(nx.pagerank(G))
# df['HubScore'] = pd.Series(nx.hits(G)[0])
# df['AuthScore'] = pd.Series(nx.hits(G)[1])

# #Train Test Seperation
# trainDf = df[~df.ManagementSalary.isnull()]
# testDf = df[df.ManagementSalary.isnull()]

# #Get Dummies
# trainDf = pd.get_dummies(trainDf, columns=['Department'])
# testDf = pd.get_dummies(testDf, columns=['Department'])

# #Align Train and Test
# trainDf, testDf = trainDf.align(testDf, join='inner', axis=1)

# #Seperate X and Y
# trainX = trainDf.drop(['ManagementSalary'], axis=1)
# train_y = trainDf.ManagementSalary.astype('int')
# testDf.drop(['ManagementSalary'], axis=1, inplace=True)

# #Train Model
# rf_clf = RandomForestClassifier(n_estimators=5000, random_state=1, n_jobs=-1, verbose=True).fit(trainX, train_y)

# #View Feature Importance
# feature_importance = pd.Series(rf_clf.feature_importances_, index=testDf.columns.tolist()).sort_values(ascending=False)[:20]
# plt.figure(figsize=(10, 8))
# sns.barplot(y = feature_importance.index, x = feature_importance.values)
# plt.show()

### Plot Function

In [89]:
# def plotFeatureSeperation(feature):

#     plt.figure()

#     sns.kdeplot(trainDf.loc[trainDf.ManagementSalary == 0, feature], label = 'target == 0')
#     sns.kdeplot(trainDf.loc[trainDf.ManagementSalary == 1, feature], label = 'target == 1')

#     plt.show();

In [173]:
def salary_predictions():
    
    df = pd.DataFrame(index=G.nodes())
    
    #Feature Extraction
    df['Department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
    df['ManagementSalary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
    df['Clustering'] = pd.Series(nx.clustering(G))
    df['Degree'] = pd.Series(G.degree())
    df['Deg_Cent'] = pd.Series(nx.degree_centrality(G))
    df['Clo_Cent'] = pd.Series(nx.closeness_centrality(G))
    df['Btn_Cent'] = pd.Series(nx.betweenness_centrality(G))
    df['PR_Scaled'] = pd.Series(nx.pagerank(G))
    df['HubScore'] = pd.Series(nx.hits(G)[0])
    df['AuthScore'] = pd.Series(nx.hits(G)[1])
    
    #Train Test Seperation
    trainDf = df[~df.ManagementSalary.isnull()]
    testDf = df[df.ManagementSalary.isnull()]
    
    #Get Dummy
    trainDf = pd.get_dummies(trainDf, columns=['Department'])
    testDf = pd.get_dummies(testDf, columns=['Department'])
    
    #Align Train/Test
    trainDf, testDf = trainDf.align(testDf, join='inner', axis=1)
    
    #Seperate X and y
    trainX = trainDf.drop(['ManagementSalary'], axis=1)
    train_y = trainDf.ManagementSalary.astype('int')
    testDf.drop(['ManagementSalary'], axis=1, inplace=True)
    
    #Train Model
    rf_clf = RandomForestClassifier(n_estimators=5000, random_state=1).fit(trainX, train_y)
    
    proba = rf_clf.predict_proba(testDf)[:, 1]
    
    submission = pd.Series(proba, index=testDf.index)
    
    return submission

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [106]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

### Feature Extraction

In [107]:
# future_connections['PrefAttach'] = [i[2] for i in nx.preferential_attachment(G, future_connections.index)]
# future_connections['CommonNeighbors'] = future_connections.index.map(lambda node: len(list(nx.common_neighbors(G, node[0], node[1]))))
# future_connections['JacdCoef'] = [item[2] for item in nx.jaccard_coefficient(G, future_connections.index)]
# future_connections['ResAlloc'] = [item[2] for item in nx.resource_allocation_index(G, future_connections.index)]
# future_connections['Ada-AdarIdx'] = [item[2] for item in nx.adamic_adar_index(G, future_connections.index)]

### Train Test Seperation

In [108]:
# trainDf = future_connections[~future_connections['Future Connection'].isnull()].copy()
# testDf = future_connections[future_connections['Future Connection'].isnull()].copy()
# testDf.drop(['Future Connection'], axis=1, inplace=True)

### Seperate X and y

In [116]:
# train_X = trainDf.drop(['Future Connection'], axis=1)
# train_y = trainDf.loc[:, 'Future Connection'].astype('int')

### Train Model

In [122]:
# logit = LogisticRegression().fit(train_X, train_y)
# submission = pd.Series(logit.predict_proba(testDf)[:, 1], index=testDf.index)

In [125]:
def new_connections_predictions():
    
    future_connections['PrefAttach'] = [i[2] for i in nx.preferential_attachment(G, future_connections.index)]
    future_connections['CommonNeighbors'] = future_connections.index.map(lambda node: len(list(nx.common_neighbors(G, node[0], node[1]))))
    future_connections['JacdCoef'] = [item[2] for item in nx.jaccard_coefficient(G, future_connections.index)]
    future_connections['ResAlloc'] = [item[2] for item in nx.resource_allocation_index(G, future_connections.index)]
    future_connections['Ada-AdarIdx'] = [item[2] for item in nx.adamic_adar_index(G, future_connections.index)]
    
    trainDf = future_connections[~future_connections['Future Connection'].isnull()].copy()
    testDf = future_connections[future_connections['Future Connection'].isnull()].copy()
    testDf.drop(['Future Connection'], axis=1, inplace=True)
    
    train_X = trainDf.drop(['Future Connection'], axis=1)
    train_y = trainDf.loc[:, 'Future Connection'].astype('int')
    
    logit = LogisticRegression().fit(train_X, train_y)
    
    submission = pd.Series(logit.predict_proba(testDf)[:, 1], index=testDf.index)
    
    return submission